### ISSUE TITLE
Remove INLINE pragma from generalCategory

### ISSUE THREAD

**Issue by mpickering** [haskell/core-libraries-committee#130]
In module GHC.Unicode.Internal.Char.UnicodeData.GeneralCategory:

generalCategory contains a huge literal string but is marked INLINE, this will duplicate the string into any use site of generalCategory. In particular `generalCategory` is used in functions like `isSpace` and the literal gets inlined into this function which makes it massive. 

My recommendation: Extract the literal into a top-level string (marked NOINLINE) and remove `INLINE` pragma from `generalCategory.

https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9958/

**Comment by treeowl** [comment]
Marking the string `NOINLINE` seems like a no-brainier. Could you explain why you want to remove the `INLINE` from the function? Where might we see advantages/disadvantages?

**Comment by phadej** [comment]
The `GHC.Unicode.Internal.Char.UnicodeData.GeneralCategory` module doesn't exist in base-4.17. it's added in May 2022.

In `base-4.17` `generalCategory` is implemented with FFI.

I'd say whether the new implementation needs tweaks to work as prior is a bug fix, which doesn't need CLC approval (but  the change of implementation probably should had a proposal).

**Comment by mpickering** [comment]
@treeowl The README states that "Changes which affect performance or laziness and similar are deemed visible. " and given the CLC approved this MR originally I think it's sensible to raise the discussion here.

**Comment by phadej** [comment]
> @treeowl The README states that "Changes which affect performance or laziness and similar are deemed visible. " and given the CLC approved this MR originally I think it's sensible to raise the discussion here.

That is fair. I'd hope for some criteria that some bugfix-y changes could be discussed just on GitLab MRs (i.e. have lighter process), but maybe it's clearer if everything goes through "proposal process", even with the cost that completely unrelated people (like me, and @treeowl) comment on the issues.

**Comment by mpickering** [comment]
I suppose the issue with making it top-level is that currently the literal is an `Addr#`.

**Comment by phadej** [comment]
> I suppose the issue with making it top-level is that currently the literal is an `Addr#`.

I was going to mention that. And if you wrap it into a `Ptr`, there will be indirection.

But AFAIK GHC can pull `Addr#` to the top-level (it does so for Typeable stuff e.g.

```haskell
-- RHS size: {terms: 1, types: 0, coercions: 0, joins: 0/0}
Data.Bifunctor.Assoc.$tcAssoc3 :: ghc-prim:GHC.Prim.Addr#
[GblId,
 Unf=Unf{Src=<vanilla>, TopLvl=True, Value=True, ConLike=True,
         WorkFree=True, Expandable=True, Guidance=IF_ARGS [] 30 0}]
Data.Bifunctor.Assoc.$tcAssoc3 = "Assoc"#
```
)
but if you add `NOINLINE`, then will it or not. IIRC the w/w will be done, and wrapper will be inlined?

So I'd say, make a PR with `-ddump-simpl` and check what the Core looks like.

**Comment by mpickering** [comment]
Here is one implementation. If the CLC would prefer a different one then it would be good for someone else to prepare a patch. https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9958/

**Comment by Bodigrim** [comment]
Dear CLC members, let's vote on the proposal to prevent inlining of a huge binary literal `bitmap#` in [`Data.Char.generalCategory`](https://ghc.gitlab.haskell.org/ghc/doc/libraries/base-4.18.0.0/Data-Char.html#v:generalCategory) as per https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9958/. @phadej kindly [checked](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9958#note_481190) that Core looks good now. 

@angerman @chshersh @hasufell @mixphix @tomjaguarpaw @parsonsmatt 

----

+1 from me.

**Comment by tomjaguarpaw** [comment]
+1

**Comment by hasufell** [comment]
+1

**Comment by angerman** [comment]
+1

**Comment by chshersh** [comment]
+1

<hr>

W.r.t. to whether such changes should be under the CLC provision or not, I'm okay with keeping the status quo. Reviewing such proposals doesn't take much time (most heavy work is done by @mpickering and @phadej here), and it's still good to share knowledge about what's going on in `base` with more people so CLC members could make more informed decisions in the future as well 👍🏻

**Comment by parsonsmatt** [comment]
Seems like a clear improvement in every way - +1 from me

**Comment by Bodigrim** [comment]
Thanks all, 6 votes in favour out of 7 are sufficient to approve.

**Comment by chshersh** [comment]
I'm trying to summarise the state of this proposal as part of my volunteering effort to track the progress of all `approved` CLC proposals.

| Field | Value |
|---|---|
| **Authors**            |  @mpickering   |
| **Status**             |  merged |
| **`base` version**     | `4.19.0.0`  |
| **Merge Request (MR)** | https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9958/ |
| **Blocked by**         | nothing |
| **CHANGELOG entry**    | present  |
| **Migration guide**    |  not needed  |

[changelog]: https://gitlab.haskell.org/ghc/ghc/-/blob/master/libraries/base/changelog.md

Please, let me know if you find any mistakes 🙂

**Comment by wismill** [comment]
It is included in `base-4.18`. It fixes #59. 

@mpickering  I think there should be no changelog entry (or merged with the one of #59 for 4.18) as the incorrect implementation it fixes was never published.

**Comment by chshersh** [comment]
@wismill Thanks for the correction! I see a changelog entry for `base-4.19`:

* https://gitlab.haskell.org/ghc/ghc/-/blob/master/libraries/base/changelog.md

But from the MR comments, I also see that this implementation was backported to GHC 9.6 (`base-4.18`).
